lib/auth/oidc_google.go

```diff
@@ -0,0 +1,214 @@
+	// all groups to work reliably.
+	credentialsParams := google.CredentialsParams{
+		Subject: connector.GetGoogleAdminEmail(),
+	}
+
+	if connector.GetGoogleTransitiveGroups() {
+		log.Debugf("Loading credentials to fetch transitive groups")
+		credentialsParams.Scopes = []string{cloudidentity.CloudIdentityGroupsReadonlyScope}
+	} else {
+		log.Debugf("Loading credentials to fetch direct groups")
+		credentialsParams.Scopes = []string{directory.AdminDirectoryGroupReadonlyScope}
+	}
+
+	credentials, err := google.CredentialsFromJSONWithParams(ctx, jsonCredentials, credentialsParams)
+	if err != nil {
+		return nil, trace.BadParameter("unable to parse google service account from %v: %v", credentialLoadingMethod, err)
+	}
+
+	return credentials, nil
+}
+
+func groupsFromGsuiteDirectory(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
+	service, err := directory.NewService(ctx, clientOptions...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var groups []string
+	err = service.Groups.List().
+		UserKey(email).
+		Pages(ctx, func(resp *directory.Groups) error {
+			if resp == nil {
+				return nil
+			}
+			for _, g := range resp.Groups {
+				if g != nil && g.Email != "" {
+					groups = append(groups, g.Email)
+				}
+			}
+			return nil
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return groups, nil
+}
+
+func groupsFromGsuiteCloudIdentity(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
+	service, err := cloudidentity.NewService(ctx, clientOptions...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var groups []string
+	// SearchTransitiveGroups takes a fixed parameter as part of the URL
+	// ("Format: `groups/{group}`, where `group` is always '-'") and a query
+	// parameter that the google API docs claim to be a CEL expression
+	// (https://opensource.google/projects/cel) that filters the results based
+	// on `member_key_id`, optionally `member_key_namespace`, and `labels`. The
+	// query parameter doesn't seem to actually be a CEL expression, and even
+	// changing the single quotes into double quotes (which is fine according to
+	// the CEL grammar) makes every API call fail with an "Unauthorized" error
+	// message.
+	//
+	// The query string was lifted directly from
+	// https://cloud.google.com/identity/docs/how-to/query-memberships#searching_for_all_group_memberships_of_a_member
+	// and some more informations on group labels can be found at
+	// https://cloud.google.com/identity/docs/groups#group_properties . The
+	// actual docs for the API call are at
+	// https://cloud.google.com/identity/docs/reference/rest/v1/groups.memberships/searchTransitiveGroups .
+	err = service.Groups.Memberships.SearchTransitiveGroups("groups/-").
+		Query(fmt.Sprintf("member_key_id == '%s' && 'cloudidentity.googleapis.com/groups.discussion_forum' in labels", email)).
```

Comment:
How does this query language work? Does it have the equivalent of parameterized queries?

I'm wondering if you can do something like SQL injection with it?

Reply:
I tried looking for a "proper" way to do this, but the only library to handle this (https://github.com/google/cel-go) is apparently only for evaluating CEL expressions in text form, not for building them; I don't think it's a problem here, because the `email` parameter we use to format the string is the `email` that came as part of the OIDC claims, so it's something that Google considers to be a valid email address - which I don't believe can include any `'`s.